app/assets/javascripts/stimulus/loaders/autoloader.js

```diff
@@ -3,19 +3,32 @@ import { Application } from "stimulus"
 const application = Application.start()
 
 function autoload() {
-  Array.from(document.querySelectorAll('[data-controller]')).forEach((element) => {
-    const controllerNames = element.attributes["data-controller"].value.split(" ")
+  for (const element of document.querySelectorAll("[data-controller]")) {
+    autoloadControllers(element)
+  }
+}
 
-    controllerNames.forEach((controllerName) => {
-      let controllerFilename = `${controllerName}_controller`
+function autoloadControllers(element) {
+  const controllerTokens = element.getAttribute("data-controller") || ""
+  const controllerNames = controllerTokens.split(/\s+/).filter(content => content.length)
 
-      import(controllerFilename).then((controllerModule) => {
-        application.register(controllerName, controllerModule.default)
-      }).catch(error => console.log(`Failed to autoload controller: ${controllerName}`))
-    })
-  })
+  for (const controllerName of controllerNames) {
+    const filename = controllerName.replace(/--/g, "/").replace(/-/g, "_")
+    import(`${filename}_controller`).then(controllerModule => {
```

Comment:
Because this code is running async on first load for 10 DOM elements with `<div data-controller="nav">` we will have 100 connect events instead of 10. To prevent it we need to track loaded controllers.

On top after `Application.start()` add following:
```js
const application = Application.start()

const loaded = {}
```

Now implement tracking of loaded controllers:
```js
import(`${filename}_controller`).then(controllerModule => {
      if(controllerName in loaded){ return }

      application.register(controllerName, controllerModule.default)
      loaded[controllerName] = true
    }).catch(error => console.log(`Failed to autoload controller: ${controllerName}`))
 ```